### PR TITLE
feat(scale): mutiple line needles

### PR DESCRIPTION
**Pull request by liamHowatt**
### Description of the feature or fix

Related issue: #5906

#### Feature

- More than one scale can have a line needle at the same time
- One scale can have multiple line needles

Scale image needles could do this already.

Doing this for line needles is nontrivial because line points are not stored in the line object. They need to be stored somewhere else. Until now, the line needle points array was `static`.

#### Changes

- Add a method to add line needles to the scale. A struct containing the line and line points gets added to the end of a linked list in scale, similar to how scale stores its scale sections.
- Change the line needle value setter to accept a line needle instead of a line. The passed line needle must come from the above mentioned line needle adder.
- Add a getter to get the line needle's underlying line object.
- Update `lv_example_scale_3`.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.10](https://github.com/szepeviktor/astyle/releases/tag/v3.4.10) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by kisvegabor** [comment]
The original implementation needs to improved for sure, but I'm little bit concerned if we really need an other type here. 

Can we just say that it's the user's responsibility to set  point array for the line before using it in `lv_scale_set_line_needle_value`? If the points are not set we can even dynamically allocate it and add an event to line which frees the points on `DELETE`.

What do you think?

**Comment by s-hadinger** [comment]
@kisvegabor I would prefer like you mention above: let the caller create the lv_line object and allocate the points.

**Comment by liamHowatt** [comment]
What are both of your thoughts on what I pushed?

We can do the caller provided points but I could not think of a clear way to document the API.
```c
/**
 * Point the needle to the corresponding value through the line
 * @param obj              pointer to a scale object
 * @param needle_line      needle_line of the scale
 * @param needle_line_points user provided uninitialized 2-array for the line points -- must not go out of scope
 * @param needle_length    length of the needle
 *                         needle_length>0 needle_length=needle_length;
 *                         needle_length<0 needle_length=radius-|needle_length|;
 * @param value            needle to point to the corresponding value
 */
void lv_scale_set_line_needle_value(lv_obj_t * obj, lv_obj_t * needle_line, lv_point_precise_t * needle_line_points,
                                    int32_t needle_length, int32_t value);
```

Is this what you had in mind?

**Comment by kisvegabor** [comment]
As `lv_obj_t * needle_line` is a real line widget, it could be used like this:
```c
static lv_point_precise_t points[2];
lv_obj_t * line = lv_line_create(scale);
lv_line_set_points(line, points, 2);

lv_scale_set_line_needle_value(scale, line, length, value);
```

**Comment by liamHowatt** [comment]
Thanks for clarifying. I feel silly needing it spelled out but what would happen inside `lv_scale_set_line_needle_value`? Would it call a new getter on the `line` to get the pointer out of the `line` or would it access the `line`'s field directly? I toyed with that earlier and the problem I ran into was that the `line` stores the array as a `const` pointer. Since the assumption is valid in this special case, can we cast away the `const`?

**Comment by kisvegabor** [comment]
Ah, that's not silly. Actually it's hard to resolve. We cannot remove the `const` from `lv_line_t` to allow using real const array. However converting the to non-const is not safe.

In lack of a better idea, I vote for the non safe way. If the array is `const` it will cause an immediate easy to debug crash. But we need to highlight the use of non-const array in the docs, in the example and the API comments too.

**Comment by liamHowatt** [comment]
I added an experimental line points mutability API. It's the maximum amount of verbosity. Let's agree how much to cut out.

In total:
- `lv_line` points mutable setter, const points getter, fallible mutable points getter, is_mutable getter, point count getter.
- `lv_scale` checks if line has mutable 2-array and uses that. Otherwise allocates one which lasts for the lifetime of the line.
- New `lv_example_scale_6` which uses 2 line needles and demonstrates the line point options.
- New tests for `lv_line` and `lv_scale`.

image

File: examples/widgets/scale/lv_example_scale_6.c [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    /**
+     * the scale will store the needle line points in the existing
+     * point array if one was set with `lv_line_set_points_mutable`.
+     * Otherwise, it will allocate the needle line points.
+     */
+
+    /* the scale will store the minute hand line points in `minute_hand_points` */
+    lv_line_set_points_mutable(minute_hand, minute_hand_points, 2); /* only needs to be set once */
```
Comment (kisvegabor) [inline examples/widgets/scale/lv_example_scale_6.c]: Why is it called in the timer if needs to be called only once?

File: src/widgets/line/lv_line.h [hunk 2]
```diff
@@ -27,9 +27,13 @@ extern "C" {
 /*Data of line*/
 typedef struct {
     lv_obj_t obj;
-    const lv_point_precise_t * point_array;     /**< Pointer to an array with the points of the line*/
+    union {
+        const lv_point_precise_t * constant;
+        lv_point_precise_t * mutable;
```
Comment (kisvegabor) [inline src/widgets/line/lv_line.h]: Note this error.

File: src/widgets/line/lv_line.h [hunk 3]
```diff
@@ -70,6 +82,35 @@ void lv_line_set_y_invert(lv_obj_t * obj, bool en);
...
+ * @return              number of points in array of points
+ */
+uint32_t lv_line_get_point_count(lv_obj_t * obj);
+
+/**
+ * Get the mutability of the stored point array pointer.
+ * @param obj           pointer to a line object
+ * @return              true: the point array pointer is mutable, false: constant
+ */
+bool lv_line_get_point_array_is_mutable(lv_obj_t * obj);
```
Comment (kisvegabor) [inline src/widgets/line/lv_line.h]: What about this?
```suggestion
bool lv_line_is_point_array_mutable(lv_obj_t * obj);
```

Having both `get` and `is` a little bit stage to me. I vote for keeping `is` as there is no direct `set` counterpart.

File: src/widgets/line/lv_line.c [hunk 4]
```diff
@@ -91,6 +89,42 @@ void lv_line_set_y_invert(lv_obj_t * obj, bool en)
...
+    return line->point_array_is_mutable;
+}
+
+lv_point_precise_t * lv_line_get_points_mutable(lv_obj_t * obj)
+{
+    LV_ASSERT_OBJ(obj, MY_CLASS);
+
+    lv_line_t * line = (lv_line_t *)obj;
+    if(!line->point_array_is_mutable) {
+        LV_LOG_ERROR("the line point array is not mutable");
```
Comment (kisvegabor) [inline src/widgets/line/lv_line.c]: It should only a warning. `ERROR`s are rather used on critical problems or crashes.

**Review by kisvegabor — COMMENTED** [review]
I like the idea of `lv_line_set_points_mutable`!

My only concern is that usually we do it the other was around. E.g.
```c
lv_label_set_text(label, buf); //buf is not const
lv_label_set_text_static(label, "hello"); //the text is constant
```

But we can keep your suggested API for backward compatibility.

File: examples/widgets/scale/lv_example_scale_6.c (hunk 1 above)
Comment (liamHowatt) [reply examples/widgets/scale/lv_example_scale_6.c]: I wanted related functionality to be nearby so they can see all the steps in one place but yes I think it's a better demonstration to call it once outside the timer 👍️

**Comment by liamHowatt** [comment]
> I like the idea of `lv_line_set_points_mutable`!
> 
> My only concern is that usually we do it the other was around. E.g.
> 
> ```c
> lv_label_set_text(label, buf); //buf is not const
> lv_label_set_text_static(label, "hello"); //the text is constant
> ```
> 
> But we can keep your suggested API for backward compatibility.

Using `_static` naming for consistency would be good but I think this is a bit different than the `set_text` cases. Technically `set_points` and `set_points_mutable` are both static setters since they don't make a copy. They technically should be called `set_points_static` and `set_points_static_mutable`.

Does that give you any ideas?

**Comment by kisvegabor** [comment]
> Does that give you any ideas?

Yes, let's leave it as it is :slightly_smiling_face:

**Review by kisvegabor — APPROVED** [review]
LGTM!

**Comment by xinsuinizhuan** [comment]
> I added an experimental line points mutability API. It's the maximum amount of verbosity. Let's agree how much to cut out.
> 
> In total:
> 
>     * `lv_line` points mutable setter, const points getter, fallible mutable points getter, is_mutable getter, point count getter.
> 
>     * `lv_scale` checks if line has mutable 2-array and uses that. Otherwise allocates one which lasts for the lifetime of the line.
> 
>     * New `lv_example_scale_6` which uses 2 line needles and demonstrates the line point storage options.
> 
>     * New tests for `lv_line` and `lv_scale`.
> 
> 
> image

where is the example? I need it.

**Comment by liamHowatt** [comment]
(see my comment in the issue https://github.com/lvgl/lvgl/issues/5972#issuecomment-2025289110)

**Review by FASTSHIFT — APPROVED** [review]

**Pull request merged**